compiler/rustc_builtin_macros/src/format.rs

```diff
@@ -160,15 +160,26 @@ fn parse_args<'a>(
 
     while p.token != token::Eof {
         if !p.eat(&token::Comma) {
-            if first {
+            let mut err = if first {
                 // After `format!(""` we always expect *only* a comma...
                 let mut err = ecx.struct_span_err(p.token.span, "expected token: `,`");
                 err.span_label(p.token.span, "expected `,`");
                 p.maybe_annotate_with_ascription(&mut err, false);
-                return Err(err);
+                err
             } else {
                 // ...after that delegate to `expect` to also include the other expected tokens.
-                let _ = p.expect(&token::Comma)?;
+                p.expect(&token::Comma).unwrap_err()
```

Comment:
I am unsure that this call of `unwrap_err` won't panic. Is there a better option?

Reply:
A better option is to always use `p.expect()`, but do
```rust
if first {
    p.expected_tokens.clear();
}
```
before that.

`expected_tokens` is private though, but you can add a public `clear_expected_tokens` method to `Parser`.

Reply:
Nice, thanks for the tip.